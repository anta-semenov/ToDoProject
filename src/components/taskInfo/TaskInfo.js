import React from 'react'
import PureRenderMixins from 'react-addons-pure-render-mixin'
import { Map, Set } from 'immutable'
import ImmutablePropTypes from 'react-immutable-proptypes'

import TaskTitle from './taskTitle/TaskTitle'
import TaskDescription from './taskDescription/TaskDescription'
import TaskContexts from './taskContexts/TaskContexts'
import TaskCalendar from './taskCalendar/TaskCalendar'
import Today from '../controls/today/Today'
import Checkbox from '../controls/checkbox/Checkbox'
import Priority from '../controls/priority/Priority'
import CloseBtn from '../controls/closeBtn/CloseBtn'

import './TaskInfo.less'

export default class TaskInfo extends React.Component {
  constructor(props) {
    super(props)
    this.shouldComponentUpdate = PureRenderMixins.shouldComponentUpdate.bind(this)
  }

  render() {
    return (
      <div className={`task-info ${this.props.id != '' ? 'is-open' : ''}` }>
        <CloseBtn appearance='task-info' onClick={this.props.onCloseClick} />
        {this.props.id != -1 ?
          <div className='task-info__content'>
            <div className='task-info__controls'>
              <Checkbox appearance='task-info' checked={this.props.completed} onClick={() => this.props.onTaskCheckboxClick(this.props.id, !this.props.completed)} />
              <Today appearance='task-info' checked={this.props.today} onClick={() => this.props.onTaskTodayClick(this.props.id, !this.props.today)}/>
              <Priority appearance='task-info' priority={this.props.priority} onClick={(priority) => this.props.onPriorityClick(this.props.id, priority)} />
            </div>
            <div className='task-info__body'>
              <div className='task-info__body-top'>
                <TaskTitle id={this.props.id} title={this.props.title} onChange={this.props.onTitleChange} />
                <TaskDescription id={this.props.id} description={this.props.description} onChange={this.props.onDescriptionChange} />
                <TaskContexts contexts={this.props.contexts} taskContexts={this.props.taskContexts} onContextClick={(context, status) => this.props.onContextClick(this.props.id, context, status)} />
                <TaskCalendar id={this.props.id} selectedDate={this.props.date} onChange={this.props.onDateChange}/>
              </div>
              <button className='task-info__delete' onClick={() => this.props.onTaskDeleteClick(this.props.id)} tabIndex='0' >Delete task</button>
            </div>
          </div> : null
        }
      </div>
    )
  }
}

TaskInfo.propTypes = {
  id: React.PropTypes.string.isRequired,
  title: React.PropTypes.string,
  completed: React.PropTypes.bool,
  today: React.PropTypes.bool,

  description: React.PropTypes.oneOfType([
    React.PropTypes.string,
    React.PropTypes.instanceOf(Map)
  ]),
  priority: React.PropTypes.string,
  date: React.PropTypes.instanceOf(Date),
  taskProject: React.PropTypes.string,
  taskContexts: React.PropTypes.instanceOf(Set),

  contexts: ImmutablePropTypes.mapOf(
    ImmutablePropTypes.contains({
      id: React.PropTypes.string.isRequired,
      title: React.PropTypes.string.isRequired
    })
  ).isRequired,

  onTaskCheckboxClick: React.PropTypes.func.isRequired,
  onTaskTodayClick: React.PropTypes.func.isRequired,
  onPriorityClick: React.PropTypes.func.isRequired,
  onTitleChange: React.PropTypes.func.isRequired,
  onDescriptionChange: React.PropTypes.func.isRequired,
  onProjectChange: React.PropTypes.func.isRequired,
  onContextClick: React.PropTypes.func.isRequired,
  onDateChange: React.PropTypes.func.isRequired,
  onTaskDeleteClick: React.PropTypes.func.isRequired,
  onCloseClick: React.PropTypes.func.isRequired
}
