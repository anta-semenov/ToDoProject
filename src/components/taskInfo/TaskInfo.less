@import '../../styles/variables.less';

.task-info {
  box-sizing: border-box;
  top: 0;
  height: 100%;
  width: 100%;
  padding: 40px 60px 40px 50px;
  margin: 0;
  overflow-y: auto;

  background-color: #ffffff;
  border-left: 1px solid @color-border-light-gray;

  @media (max-width: 1270px) {
    box-shadow: -2px 0 30px 0 fade(#000000, 6%), 0 0 10px 0 fade(#000000, 4%);
  }
}

.task-info__content {
  position: relative;
  min-width: @size-width-task-info - 110;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
}

.task-info__controls {
  position: relative;
  margin: 4px 14px 0 0;
  flex-shrink: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
}
.task-info__body {
  position: relative;
  min-height: 100%;
  margin: 0;
  flex-grow: 1;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: space-between;
}
.task-info__body-top {
  position: relative;
  width: 100%;
}

.task-info__delete {
  position: relative;
  z-index: 1;
  display: block;
  margin: 40px 0 0 4px;
  padding: 6px 12px 7px 12px;
  cursor: pointer;

  text-transform: capitalize;
  font-size: @font-size-basic;
  font-weight: @font-weight-medium;
  color: @color-red;
  border: 1px solid fade(@color-red, 40%);
  border-radius: 2px;
  background-color: #ffffff;

  transition: @transition-basic;

  &:hover,
  &:focus {
    border-color: @color-red;
    background-color: fade(desaturate(@color-red, 35%), 5%);
    outline: none;
  }
  &:active {
    border-color: @color-red;
    background-color: @color-red;
    color: #ffffff;
  }
}
.task-info__delete-text {
  position: relative;
  z-index: 2;
}
