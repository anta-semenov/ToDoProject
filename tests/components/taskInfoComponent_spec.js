import React from 'react'
import { shallow } from 'enzyme'
import TaskInfo from '../../src/components/taskInfo/TaskInfo'

describe('TaskInfo component', () => {
  const props = {
    onTaskCheckboxClick: () => {},
    onTaskTodayClick: () => {},
    onPriorityClick: () => {},
    onTitleChange: () => {},
    onDescriptionChange: () => {},
    onProjectChange: () => {},
    onContextClick: () => {},
    onDateChange: () => {},
    onCloseClick: () => {},
    onTaskSomedayClick: () => {},
    onTaskDeleteClick: () => {}
  }
  describe('Delete Task', () => {
    test('Should render delete task button', () => {
      const taskInfoComponent = shallow(<TaskInfo id={'b41sogy3s0o2'} {...props} />)
      expect(taskInfoComponent).toMatchSnapshot()
    })
    test('Should handle click on delete task button', () => {
      const onTaskDeleteClick = jest.fn()
      const taskInfoComponent = shallow(
        <TaskInfo id={'b41sogy3s0o2'} {...props} onTaskDeleteClick={onTaskDeleteClick} />
      )
      taskInfoComponent.find('.task-info__delete').simulate('click')

      expect(onTaskDeleteClick).toHaveBeenCalled()
    })
  })
})
