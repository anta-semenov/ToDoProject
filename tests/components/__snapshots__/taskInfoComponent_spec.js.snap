// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskInfo component Delete Task Should render delete task button 1`] = `
<div
  className="task-info"
>
  <CloseBtn
    appearance="task-info"
    onClick={[Function]}
  />
  <div
    className="task-info__content"
  >
    <div
      className="task-info__controls"
    >
      <Checkbox
        appearance="task-info"
        checked={false}
        dimmed={false}
        onChange={[Function]}
      />
      <Today
        appearance="task-info"
        onClick={[Function]}
      />
      <Someday
        appearance="task-info"
        onClick={[Function]}
      />
      <Priority
        appearance="task-info"
        onClick={[Function]}
      />
    </div>
    <div
      className="task-info__body"
    >
      <div
        className="task-info__body-top"
      >
        <TaskTitle
          id="b41sogy3s0o2"
          onBlur={[Function]}
          onChange={[Function]}
        />
        <TaskDescription
          id="b41sogy3s0o2"
          onBlur={[Function]}
          onChange={[Function]}
        />
        <TaskProject
          onProjectChange={[Function]}
        />
        <TaskContexts
          onContextClick={[Function]}
        />
        <TaskCalendar
          id="b41sogy3s0o2"
          onChange={[Function]}
        />
        <TaskRepeat
          onRepeatChange={[Function]}
        />
      </div>
      <button
        className="task-info__delete"
        onClick={[Function]}
        tabIndex="0"
      >
        Delete task
      </button>
    </div>
  </div>
</div>
`;
